### PR TITLE
CSV: store dialect options and link to source

### PR DESCRIPTION
**Pull request by vhf**
#63

### CODE REVIEW

File: e2e-tests/csv-source/upload.hydra [hunk 1]
```diff
@@ -41,11 +41,18 @@ With Class cc:CubeProject {
                     ```
                     A,B,C
                     1,2,test
+                    1,2,test
+                    1,2,test
+                    1,2,test
```
Comment (vhf) [inline e2e-tests/csv-source/upload.hydra:46]: Two lines were not enough for csv-sniff to do its magic (Error: No newline characters found in your file... at CSVSniffer.sniff (/app/node_modules/csv-sniffer/index.js:631:10))

File: e2e-tests/csv-source/upload.hydra [hunk 2]
```diff
@@ -41,11 +41,18 @@ With Class cc:CubeProject {
...
                     ```
                 } => {
                     Expect Status 201
                     Expect Type <https://cube-creator.zazuko.com/vocab#CSVSource>
                     Expect Property <http://schema.org/associatedMedia>
+                    Expect Property csvw:quoteChar
+                    Expect Property csvw:delimiter
+                    Expect Property csvw:header
+                    Expect Property csvw:headerRowCount
```
Comment (tpluscode) [inline e2e-tests/csv-source/upload.hydra:55]: Please package these in a `csvw:Dialect`
```suggestion
                    Expect Property csvw:dialect {
                        Expect Property csvw:quoteChar
                        Expect Property csvw:delimiter
                        Expect Property csvw:header
                        Expect Property csvw:headerRowCount
                    }
```

File: e2e-tests/csv-source/upload.hydra (hunk 1 above)
Comment (tpluscode) [reply e2e-tests/csv-source/upload.hydra:46]: Curious. Maybe the case of failed sniffing could make for a mocha test to check that it does not fail and a resource is indeed added to the store (can use `TestResourceStore` + `to.matchSnapshot(this)`

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: apis/core/lib/domain/csv-source/upload.ts [hunk 3]
```diff
@@ -37,6 +40,22 @@ export async function uploadFile({
...
+      .addOut(csvw.dialect, csvDialect => {
+        csvDialect.addOut(csvw.quoteChar, dialect.quote)
+          .addOut(csvw.delimiter, dialect.delimiter)
+          .addOut(csvw.header, true)
+          .addOut(csvw.headerRowCount, header.length)
+      })
+  } catch (err) {
+    error(err)
+  }
+
```
Comment (lucafurrer) [inline apis/core/lib/domain/csv-source/upload.ts:58]: What do you think, should we extract this to an other file? After the discussion of today, I think we will use this logic soon from an other place as well?

Comment (tpluscode) [reply apis/core/lib/domain/csv-source/upload.ts:58]: We will extract this as soon as we need IMO :)

File: apis/core/lib/domain/csv-source/upload.ts [hunk 4]
```diff
@@ -37,6 +40,22 @@ export async function uploadFile({
...
     })
 
+  try {
+    const fileStream = await loadFile(key) as Readable
+    const head = await loadFileHeadString(fileStream)
+    const { dialect, header } = await sniffParse(head)
+
+    csvSource
+      .addOut(csvw.dialect, csvDialect => {
+        csvDialect.addOut(csvw.quoteChar, dialect.quote)
```
[the remark this answers is not in the document]
Comment (lucafurrer) [reply apis/core/lib/domain/csv-source/upload.ts:50]: This syntac is confusiong me. 
Where will the namedNode dialect be attached. To the empty node? to  csvw.quoteChar?
Do you think you can write the resovling triples?

File: e2e-tests/csv-source/upload.hydra [hunk 5]
```diff
@@ -57,18 +88,21 @@ With Class cc:CSVSource {
...
+        Invoke {
             Accept "application/ld+json"
         } => {
             Expect Status 200
             Expect Header Content-Type "application/ld+json; charset=utf-8"
             Expect Type <https://cube-creator.zazuko.com/vocab#CSVSource>
+            Expect Property csvw:quoteChar
+            Expect Property csvw:delimiter
+            Expect Property csvw:header
+            Expect Property csvw:headerRowCount
```
Comment (lucafurrer) [inline e2e-tests/csv-source/upload.hydra:105]: @tpluscode Can we omit the Expect Property csvw:dialect here?

File: apis/core/lib/domain/csv-source/upload.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tpluscode) [reply apis/core/lib/domain/csv-source/upload.ts:50]: Oh sorry, my mistake on the review suggestion

Comment (tpluscode) [inline apis/core/lib/domain/csv-source/upload.ts:50]: One last thing: please make the dialect a named node:

```suggestion
      .addOut(csvw.dialect, id.dialect(csvSource), csvDialect => {
        csvDialect.addOut(csvw.quoteChar, dialect.quote)
```

by appending /dialect to source's URI in id.dialect

File: e2e-tests/csv-source/upload.hydra (hunk 5 above)
Comment (tpluscode) [reply e2e-tests/csv-source/upload.hydra:105]: No, these lines will always fail.

**Review by tpluscode — APPROVED** [review]

**Pull request merged**